c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,170 @@
+use backtrace::Backtrace;
+use log::warn;
+use rustc_span::{Span, DUMMY_SP};
+use std::any::Any;
+use std::cell::Cell;
+use std::fmt::Write as _;
+use std::panic::PanicInfo;
+
+/// Detailed information about a panic.
+#[derive(Clone, Debug)]
+pub struct PanicDetail {
+    msg: String,
+    loc: Option<String>,
+    relevant_loc: Option<String>,
+    backtrace: Option<Backtrace>,
+    sp: Span,
+}
+
+impl PanicDetail {
+    /// Create a new `PanicDetail` containing only a message, with no location or backtrace
+    /// information.
+    pub fn new(msg: String) -> PanicDetail {
+        PanicDetail {
+            msg,
+            loc: None,
+            relevant_loc: None,
+            backtrace: None,
+            sp: DUMMY_SP,
+        }
+    }
+
+    /// Returns `true` if this `PanicDetail` contains a backtrace.
+    pub fn has_backtrace(&self) -> bool {
+        self.backtrace.is_some()
+    }
+
+    /// Return a short (usually one-line) description of this panic.
+    pub fn to_string_short(&self) -> String {
+        let loc_str = self
+            .relevant_loc
+            .as_ref()
+            .or(self.loc.as_ref())
+            .map_or("[unknown]", |s| &*s);
+        format!("{}: {}", loc_str, self.msg.trim())
+    }
+
+    /// Return a full description of this panic, including a complete backtrace if available.
+    pub fn to_string_full(&self) -> String {
+        let mut s = String::new();
+        let loc_str = self.loc.as_ref().map_or("[unknown]", |s| &*s);
+        writeln!(s, "panic at {}: {}", loc_str, self.msg).unwrap();
+        if let Some(ref relevant_loc) = self.relevant_loc {
+            writeln!(s, "related location: {}", relevant_loc).unwrap();
+        }
+        if !self.sp.is_dummy() {
+            writeln!(s, "source location: {:?}", self.sp).unwrap();
+        }
+        if let Some(ref bt) = self.backtrace {
+            writeln!(s, "{:?}", bt).unwrap();
+        }
+        s
+    }
+}
+
+thread_local! {
+    static CURRENT_PANIC_DETAIL: Cell<Option<PanicDetail>> = Cell::new(None);
+}
+
+/// Panic hook for use with [`std::panic::set_hook`].  This builds a `PanicDetail` for each panic
+/// and stores it for later retrieval by [`take_current`].
+pub fn panic_hook(info: &PanicInfo) {
+    let bt = Backtrace::new();
+    let detail = PanicDetail {
+        msg: panic_to_string(info.payload()),
+        loc: info.location().map(|l| l.to_string()),
```

Comment:
I think it's fine to do `info.location().unwrap()`.  Currently, it always returns `Some`, and `std` `.unwrap()`s, too.  That would simplify a bunch of the logic.

Reply:
This change only simplifies this single line (`loc` still needs to be `Option` since `PanicDetail::new` doesn't provide a location), and it would make us reliant on an implementation detail of the standard library, so I don't think it's worth it.